compiler/codegen/symbol.cpp

```diff
@@ -2165,6 +2172,114 @@ void FnSymbol::codegenPrototype() {
   return;
 }
 
+namespace {
+  /* This little analysis helps with llvm.loop.parallel_accesses
+     metadata.
+
+     It identifies references that are defined inside an order-independent
+     loop. If it can show that these references refer to something outside
+     of any order-independent loop in the function, it marks them with
+     FLAG_POINTS_NON_STACK.
+
+     The goal here is to avoid hinting llvm.loop.parallel_accesses
+     for load/store to temporary variables that are declared inside
+     of an order independent loop. The reason for that is that the LLVM
+     optimizers do not yet handle vectorizing allocas. By not marking
+     load/store with allocas for variables within the loop as
+     parallel_accesses, vectorization will be inhibited unless these
+     are all promoted to registers (or the LLVM vectorization pass can
+     prove it is OK for some other reason).
+   */
+  struct MarkNonStackVisitor : public AstVisitorTraverse {
+    LoopStmt* outerMostOrderIndependentLoop;
+    MarkNonStackVisitor() : outerMostOrderIndependentLoop(NULL) { }
+    void handleLoopStmt(LoopStmt* loop);
+    bool exprPointsToNonStack(Expr* e);
+    bool enterCallExpr(CallExpr* call);
+    bool enterWhileDoStmt(WhileDoStmt* loop);
+    bool enterDoWhileStmt(DoWhileStmt* loop);
+    bool enterCForLoop(CForLoop* loop);
+    bool enterForLoop(ForLoop* loop);
+  };
+}
+
+void MarkNonStackVisitor::handleLoopStmt(LoopStmt* loop) {
+  if (loop->isOrderIndependent() && outerMostOrderIndependentLoop == NULL) {
+    outerMostOrderIndependentLoop = loop;
+  }
+}
+
+bool MarkNonStackVisitor::exprPointsToNonStack(Expr* e) {
+  if (SymExpr* se = toSymExpr(e)) {
+    Symbol* sym = se->symbol();
+
+    if (sym->hasFlag(FLAG_POINTS_NON_STACK))
+      return true; // already marked
+
+    if (outerMostOrderIndependentLoop != NULL &&
+        !outerMostOrderIndependentLoop->contains(sym->defPoint)) {
+      // defined outside the outermost independent loop
+      // (could be a global variable or argument, too)
+      return true;
```

Comment:
This is part of my confusion about this "non-stack" business. Won't this return true for any scalar local variable (that is in stack) in the function, if it is outside any order independent loop? Maybe the naming is what's confusing me.

I was kind of thinking that what you mean as stack in code may refer to stack of nested loops, but it appears that you are referring to stack as we know it.

Reply:
Name change should resolve this.